components/webdriver_server/actions.rs

```diff
@@ -317,5 +358,148 @@ impl Handler {
+                        Some((x_element, y_element)) => (x_element, y_element),
+                        None => return Err(ErrorStatus::UnknownError),
+                    },
+                    Err(_) => return Err(ErrorStatus::UnknownError),
+                }
+            },
+        };
+
+        let (sender, receiver) = ipc::channel().unwrap();
+        let cmd_msg = WebDriverCommandMsg::GetWindowSize(
+            self.session.as_ref().unwrap().top_level_browsing_context_id,
+            sender,
+        );
+        self.constellation_chan
+            .send(ConstellationMsg::WebDriverCommand(cmd_msg))
+            .unwrap();
+
+        let viewport = receiver.recv().unwrap().initial_viewport;
+        if x < 0 || x as f32 > viewport.width || y < 0 || y as f32 > viewport.height {
+            return Err(ErrorStatus::MoveTargetOutOfBounds);
+        }
+
+        let duration = match action.duration {
+            Some(duration) => duration,
+            None => tick_duration,
+        };
+
+        thread::sleep(Duration::from_millis(17));
+
+        self.perform_pointer_move(source_id, duration, start_x, start_y, x, y, &tick_start)?;
+
+        Ok(())
+    }
+
+    fn perform_pointer_move(
+        &mut self,
+        source_id: &str,
+        duration: u64,
+        start_x: i64,
+        start_y: i64,
+        target_x: i64,
+        target_y: i64,
+        tick_start: &Instant,
+    ) -> Result<(), ErrorStatus> {
+        let time_delta = tick_start.elapsed().as_millis();
+
+        let duration_ratio = if duration > 0 {
+            time_delta as f64 / duration as f64
+        } else {
+            1.0
+        };
+
+        let last = if 1.0 - duration_ratio < 0.001 {
+            true
+        } else {
+            false
+        };
+
+        let (x, y) = if last {
+            (target_x, target_y)
+        } else {
+            (
+                (duration_ratio * (target_x - start_x) as f64) as i64 + start_x,
+                (duration_ratio * (target_y - start_y) as f64) as i64 + start_y,
+            )
+        };
+
+        let pointer_input_state = match self
+            .session
+            .as_mut()
+            .unwrap()
+            .input_state_table
+            .get_mut(source_id)
+            .unwrap()
+        {
+            InputSourceState::Null => unreachable!(),
+            InputSourceState::Key(_) => unreachable!(),
+            InputSourceState::Pointer(pointer_input_state) => pointer_input_state,
+        };
+
+        let current_x = pointer_input_state.x;
+        let current_y = pointer_input_state.y;
+
+        if x != current_x as i64 || y != current_y as i64 {
+            let cmd_msg = WebDriverCommandMsg::MouseMoveAction(x as f32, y as f32);
+            self.constellation_chan
+                .send(ConstellationMsg::WebDriverCommand(cmd_msg))
+                .unwrap();
+        }
+
+        if last {
+            return Ok(());
+        }
+
+        thread::sleep(Duration::from_millis(17));
```

Comment:
The current implementation is synchronous, so we don't match the constraints described by https://w3c.github.io/webdriver/#issue-container-generatedID-46. I think we should address that spawning a thread that is responsible for sending the remaining webdriver commands.

Reply:
I'm not sure how I could do this while also allowing for errors triggered while performing the move to be propagated back to the main thread.

Reply:
Since `perform_move_action` needs access to `self`, we'll also need to synchronize its shared usage, is that right?

Reply:
I recommend making perform_move_action a free function that accepts all of the state it requires as arguments. Can the input_state_table entry for a particular source id change? If not, we can pass the pointer input state and the constellation channel. If it can, then we can store the input state table inside of an Arc<Mutex> and pass that along with the constellation channel.

Reply:
Running `perform_move_action` in a separate thread interferes with the `ElementClick` command, where we need to dispatch a `pointerMove`, a `pointerDown` and then a `pointerUp` action, in the proper order: https://w3c.github.io/webdriver/#element-click (steps 8.10 - 8.13). The `pointerMove` action might happen after the `pointerDown` and `pointerUp` actions, which is not desirable.